### PR TITLE
FLE Azure KMS dotnet

### PR DESCRIPTION
**Pull request by terakilobyte**
### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-13037

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodborg-staging.corp.mongodb.com/154d508/drivers/docsworker-xlarge/DOCSP-13037/security/client-side-field-level-encryption-local-key-to-kms#specify-the-azure-credentials

### CODE REVIEW

**Comment by terakilobyte** [comment]
@DmitryLukyanov Does this look good?

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 1]
```diff
@@ -126,6 +126,31 @@ content: |
...
+
+           var azureTenantId = Environment.GetEnvironmentVariable("FLE_AZURE_TENANT_ID");
+           var azureClientId = Environment.GetEnvironmentVariable("FLE_AZURE_CLIENT_ID");
+           var azureClientSecret = Environment.GetEnvironmentVariable("FLE_AZURE_CLIENT_SECRET");
+           var azureKmsOptions = new Dictionary<string, object>
+           {
+               { "tenantId", azureTenantId },
+               { "clientId", azureClientId },
+               { "clientSecret", azureClientSecret },
+
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:145]: blank line. Check the code in the patch which I sent you, I think I fixed it there

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 2]
```diff
@@ -126,6 +126,31 @@ content: |
...
+        .. code-block:: chsarp
+
+           var kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();
+
+           var azureTenantId = Environment.GetEnvironmentVariable("FLE_AZURE_TENANT_ID");
+           var azureClientId = Environment.GetEnvironmentVariable("FLE_AZURE_CLIENT_ID");
+           var azureClientSecret = Environment.GetEnvironmentVariable("FLE_AZURE_CLIENT_SECRET");
+           var azureKmsOptions = new Dictionary<string, object>
+           {
+               { "tenantId", azureTenantId },
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:142]: As I understand, it's better to mention all available options(with possible description in comments). See here: https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#kmsproviders. I also included it in the patch

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 3]
```diff
@@ -230,6 +255,31 @@ content: |
...
+        .. code-block:: csharp
+
+           var clientEncryption = new ClientEncryption(clientEncryptionOptions);
+           var azureKeyName = Environment.GetEnvironmentVariable("FLE_AZURE_KEY_NAME");
+           var azureKeyVaultEndpoint = Environment.GetEnvironmentVariable("FLE_AZURE_KEYVAULT_ENDPOINT"); // typically <azureKeyName>.vault.azure.net
+           var dataKeyOptions = new DataKeyOptions(
+               masterKey: new BsonDocument
+               {
+                   { "keyName", azureKeyName },
+                   { "keyVaultEndpoint", azureKeyVaultEndpoint }
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:270]: `keyVersion` is missed. See here: https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#masterkey

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 4]
```diff
@@ -126,6 +126,31 @@ content: |
                   "clientSecret": "<Azure client secret>"
               }
            }
+
+     .. tab::
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:130]: I have no idea about what formatting is expected here :)

**Comment by terakilobyte** [comment]
@DmitryLukyanov I've updated it to be inline with the code now.

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 5]
```diff
@@ -230,6 +242,39 @@ content: |
            To use the Azure Key Vault, you must use `pymongocrypt <https://pypi.org/project/pymongocrypt/>`__
            version 1.1 or later in your application's environment.
 
+     .. tab::
+        :tabid: csharp
+
+        .. code-block:: csharp
+
+           var keyVaultClient = new MongoClient(_connectionString);
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:250]: as I understand this constant is not mentioned anywhere in this doc? If so, I would replace it on the string value

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 6]
```diff
@@ -230,6 +242,39 @@ content: |
...
+           var clientEncryptionOptions = new ClientEncryptionOptions(
+               keyVaultClient: keyVaultClient,
+               keyVaultNamespace: _keyVaultNamespace,
```
Comment (DmitryLukyanov) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:253]: same as above

**Review by DmitryLukyanov — APPROVED** [review]
Minor comment and the LGTM

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 7]
```diff
@@ -99,16 +99,9 @@ content: |
...
 
-           // TODO: verify correctness
-           kms_providers = {
-             azure: {
-               tenantId: "<Azure account organization>",
-               clientId: "<Azure client ID>",
-               clientSecret: "<Azure client secret>"
-             }
-           }
+           Node does not currently support Azure as a remote KMS
```
Comment (ccho-mongodb) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:104]:
```suggestion
           The Node.js driver does not currently support for Azure KMS.
```

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml [hunk 8]
```diff
@@ -230,6 +242,39 @@ content: |
...
+               });
+
+           var dataKeyId = clientEncryption.CreateDataKey("azure", dataKeyOptions, CancellationToken.None);
+           Console.WriteLine($"Azure DataKeyId [UUID]: {dataKeyId}");
+           var dataKeyIdBase64 = Convert.ToBase64String(GuidConverter.ToBytes(dataKeyId, GuidRepresentation.Standard));
+           Console.WriteLine($"Azure DataKeyId [base64]: {dataKeyIdBase64}");
+
+        .. note::
+
+           To use the AWS Key Vault, you must use
```
Comment (ccho-mongodb) [inline source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:275]: Nitpick: I wonder if we can avoid "To use..., ... use ...".
How about "To access the AWS Key Vault, you must use...

**Review by ccho-mongodb — APPROVED** [review]
A couple minor items, otherwise looks good.

File: source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml (hunk 8 above)
Comment (terakilobyte) [reply source/includes/steps-fle-convert-to-a-remote-master-key-azure.yaml:275]: I think this is good and will address it when I do the GCP ticket.

**Pull request merged**